### DISCUSSION TITLE
Tesseract - Running in Browser Console

### DISCUSSION THREAD

**Discussion by DigitalGyspy** [naptha/tesseract.js#919] [General]
Sorry, I know this is not an issue at all, so I apologise for posting here.

I am trying to use Tesseract within the browser console. This is for a site for which is not my own so I cannot embed external scripts for security reasons, hence why I run my code in the browser console. However, it is only for personal use and for intention of reading data from a canvas image.

I embedded the JS from 'https://cdn.jsdelivr.net/npm/tesseract.js@5.0.5/dist/tesseract.min.js' and then ran the following from the examples page of Tesseract GitHub pages:

(async () => {
const worker = await createWorker('eng');
const ret = await worker.recognize('https://tesseract.projectnaptha.com/img/eng_bw.png');
console.log(ret.data.text);
await worker.terminate();
})();

Which produces the following error (also because of security issues.): 'VM11097:6 Uncaught (in promise) Uncaught NetworkError: Failed to execute 'importScripts' on 'WorkerGlobalScope': The script at 'https://cdn.jsdelivr.net/npm/tesseract.js@v5.0.5/dist/worker.min.js' failed to load.'

Is it possible to compile the Tesseract JS files into a single block of code? Essentially, I want to bypass anything which involves referencing an external source.

Many thanks, greatly appreciated.

**Comment by Balearica** [comment]
While I can't say that bundling Tesseract.js into a single file is impossible (somebody could probably figure it out with enough time), it is not possible with our current projects structure/build system, and I cannot imagine that making it work would be easier than finding alternative ways to accomplish whatever you're trying to do. 

Tesseract.js requires 4 files in total: the main thread scripts (`tesseract.min.js`), the worker thread scripts (`worker.min.js`), the webassembly build of Tesseract (`tesseract-core-simd.wasm.js`), and the language data (`eng.traineddata.gz` in this case).  Therefore, you would need to figure out how to embed all 4 files in a single file.

**Comment by slavistan** [comment]
@DigitalGyspy Were you able to make any progress? I'm facing a similar problem, wanting to create a self-contained library that uses tesseract.js under the hood. My application has to work without internet connection and thus can't access any external systems. Everything has to be bundled together and I can't figure out how to get around the IndexedDB setup for the cache directory storing the _traineddata_ files.

**Reply by Balearica** [reply]
The challenge with implementing what the original post is describing comes with bundling together a bunch of discrete files, including multiple JavaScript entry points that run on different threads.  Off-the-shelf bundlers do not do this, so you would need to develop something custom.

That is different from wanting to provide all files locally, which should be perfectly doable.  You can disable caching to IndexedDB by setting `cacheMethod: 'none'` as described in the [docs](https://github.com/naptha/tesseract.js/blob/master/docs/api.md).  If you think what you are trying to do is qualitatively different from what the original post is describing, feel free to open a new discussion.

**Comment by slavistan** [comment]
@DigitalGyspy I've managed to create a self-contained bundle usable without need for access to external systems. You can use this as a starting point for yourself. 
- https://github.com/slavistan/ocr-tesseractjs-selfcontained
- https://www.npmjs.com/package/ocr-tesseractjs-selfcontained
